src/librustc_mir/build/scope.rs

```diff
@@ -510,7 +545,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             resumeblk
         };
 
-        for scope in scopes.iter_mut().filter(|s| !s.drops.is_empty() || s.free.is_some()) {
+        for scope in scopes.iter_mut().filter(|s| s.needs_cleanup) {
```

Comment:
Maybe I should put this in the comment, but `StorageDead` doesn't go in the *unwind* path (which is what is being dealt with here, and everywhere else using `needs_cleanup`).

Having it in the unwind path would pollute functions with landing pads, and I've checked `clang` and it doesn't emit `llvm.lifetime` in landing pads.
Also, `llvm.lifetime` in a landing pad might even cause problems, as it's not being tested AFAICT and could interfere with the unwinding process, but I haven't tried it.